lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -427,24 +427,35 @@ struct RvalueExprVisitor {
   Value visit(const slang::ast::ElementSelectExpression &expr) {
     auto type = context.convertType(*expr.type);
     auto value = context.convertRvalueExpression(expr.value());
+    if (!type || !value)
+      return {};
+    if (auto *constValue = expr.selector().constant) {
+      auto lowBit = constValue->integer().as<uint32_t>().value();
+      return builder.create<moore::ExtractOp>(loc, type, value, lowBit);
+    }
     auto lowBit = context.convertRvalueExpression(expr.selector());
-
-    if (!type || !value || !lowBit)
+    if (!lowBit)
       return {};
-    return builder.create<moore::ExtractOp>(loc, type, value, lowBit);
+    return builder.create<moore::DynExtractOp>(loc, type, value, lowBit);
   }
 
   // Handle range bits selections.
   Value visit(const slang::ast::RangeSelectExpression &expr) {
     auto type = context.convertType(*expr.type);
     auto value = context.convertRvalueExpression(expr.value());
-    Value lowBit;
+    if (!type || !value)
+      return {};
+
+    Value dynLowBit;
+    uint32_t constLowBit;
+    auto *leftConst = expr.left().constant;
+    auto *rightConst = expr.right().constant;
     if (expr.getSelectionKind() == slang::ast::RangeSelectionKind::Simple) {
-      if (expr.left().constant && expr.right().constant) {
-        auto lhs = expr.left().constant->integer().as<uint64_t>().value();
-        auto rhs = expr.right().constant->integer().as<uint64_t>().value();
-        lowBit = lhs < rhs ? context.convertRvalueExpression(expr.left())
-                           : context.convertRvalueExpression(expr.right());
+      if (leftConst && rightConst) {
+        // Estimate whether is big endian or little endian.
+        auto lhs = leftConst->integer().as<uint32_t>().value();
+        auto rhs = rightConst->integer().as<uint32_t>().value();
+        constLowBit = lhs < rhs ? lhs : rhs;
```

Comment:
Can you add the same check here, to make sure that we at least error out if the constant is too large to fit into 32 bits?

Reply:
As above mentioned, slang will truncate the integer literal that overflows 32 bits, so we also need to add this check :thinking:?

Reply:
Nice find! :smile: